### ISSUE TITLE
Failure to connect to Kafka broker

### ISSUE THREAD

**Issue by gmocklyt** [eclipse-ditto/ditto#1977]
Hello,

I am trying to use ditto with Kafka to communicate with devices but cannot manage to have a stable connection.
The observed behaviour is that the connection will initially success but after few seconds will report the following status then tries to reconnect in a loop.
```
{
  "type": "connectivity.responses:retrieveConnectionStatus",
  "status": 200,
  "connectionId": "91a5ef28-1985-4104-9030-6077e10a2405",
  "connectionStatus": "open",
  "liveStatus": "misconfigured",
  "recoveryStatus": "ongoing",
  "clientStatus": [
    {
      "type": "client",
      "client": "2ee793a7259f",
      "status": "misconfigured",
      "recoveryStatus": "ongoing",
      "statusDetails": "Unexpected consumer failure. - cause <KafkaConnectionFailed>: Can't establish connection with kafkaBroker after 1 attempts. At 2024-07-02T10:05:15.136905305Z",
      "inStateSince": "2024-07-02T10:05:02.098167114Z"
    }
  ],
  "sourceStatus": [
    {
      "type": "source",
      "client": "2ee793a7259f",
      "address": "connectInput",
      "status": "closed",
      "statusDetails": "Closed because of unknown/failure/misconfiguration status in client."
    }
  ],
  "targetStatus": [
    {
      "type": "target",
      "client": "2ee793a7259f",
      "address": "connectThings",
      "status": "closed",
      "statusDetails": "Closed because of unknown/failure/misconfiguration status in client."
    }
  ]
}
```

Also the connectivity logs are filled with this error message:
```
2024-07-02 12:07:04 2024-07-02 12:07:04,526 WARN  [][] o.a.k.c.NetworkClient  - [Consumer clientId=91a5ef28-1985-4104-9030-6077e10a2405-consumer, groupId=91a5ef28-1985-4104-9030-6077e10a2405] Connection to node 0 (localhost/127.0.0.1:9094) could not be established. Broker may not be available.
```
This message puzzles me as it looks like the connectivity services tries to connect on localhost while the Kafka broker is on another container.

My deployment uses the docker compose configuration in `deployment/docker/docker-compose.yml` to which I added a container with the Kafka broker as such:
```
services:
  kafka:
    image: bitnami/kafka:latest
    networks:
      default:
        aliases:
          - kafka
    ports:
      - 9094:9094
    environment:
      - KAFKA_CFG_NODE_ID=0
      - KAFKA_CFG_PROCESS_ROLES=controller,broker
      - KAFKA_CFG_LISTENERS=PLAINTEXT://:9092,CONTROLLER://:9093,EXTERNAL://:9094
      - KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://kafka:9092,EXTERNAL://localhost:9094
      - KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT,EXTERNAL:PLAINTEXT
      - KAFKA_CFG_CONTROLLER_QUORUM_VOTERS=0@kafka:9093
      - KAFKA_CFG_CONTROLLER_LISTENER_NAMES=CONTROLLER
```

My connection configuration is the following, adapted from the default Kafka template to use additional policies and provide the address of the broker container:
```
{
  "id": "91a5ef28-1985-4104-9030-6077e10a2405",
  "name": "Kafka 2",
  "connectionType": "kafka",
  "connectionStatus": "open",
  "uri": "tcp://kafka:9094",
  "sources": [
    {
      "addresses": [
        "connectInput"
      ],
      "consumerCount": 1,
      "qos": 1,
      "authorizationContext": [
        "ditto:inbound-auth-subject",
        "connect:TAF-context"
      ],
      "enforcement": {
        "input": "{{ header:device_id }}",
        "filters": [
          "{{ entity:id }}"
        ]
      },
      "acknowledgementRequests": {
        "includes": []
      },
      "headerMapping": {},
      "payloadMapping": [
        "Ditto"
      ],
      "replyTarget": {
        "address": "theReplyTopic",
        "headerMapping": {},
        "expectedResponseTypes": [
          "response",
          "error",
          "nack"
        ],
        "enabled": true
      }
    }
  ],
  "targets": [
    {
      "address": "connectThings",
      "topics": [
        "_/_/things/twin/events",
        "_/_/things/live/messages"
      ],
      "authorizationContext": [
        "ditto:outbound-auth-subject",
        "connect:TAF-context"
      ],
      "headerMapping": {}
    }
  ],
  "clientCount": 1,
  "failoverEnabled": true,
  "validateCertificates": true,
  "processorPoolSize": 1,
  "specificConfig": {
    "saslMechanism": "plain",
    "bootstrapServers": "kafka:9094"
  },
  "tags": []
}
```

Do you have any insight on this issue ?

**Comment by thjaeckle** [comment]
@gmocklyt I might have an idea ..
How did you configure Kafka?
Because Ditto will contact the Kafka bootstrap server at `kafka` hostname - and the bootstrap server then responds with the actual hostnames / IPs to connect to.  
I assume that Kafka is configured to use hostname `localhost` as "advertised listener" - and responds with `localhost` - to which Ditto however can of course not connect.

So TL;DR: adjust the "advertised listener" setting in your Kafka.

**Comment by gmocklyt** [comment]
You're right, I mixed up the port intended to be exposed to my host and the one available to the docker network.

Thank you very much for taking time to answer this, I'll close the issue
